### PR TITLE
[Logs SDK] Send resource once to processor and exporter, and not for every event. 

### PR DESCRIPTION
**Pull request by lalitb**
Fixes  Part of #1124 (for logs)
Design discussion issue (if applicable) #

## Changes
1. Send resource once to processor at the startup:
    a. Add `LogExporter::set_resource(&resource)` method. 
    b. Add `LogProcessor::set_resource(&resource)` method, which invokes `set_resource` on all the exporters.
    c. The LogExporter would store the resource in original or serialized format, and merge it with each event during export.
    d. The LoggerProvider invokes `set_resource` on all the processors during its initialization.
    e. Remove `resource` from the `LogData` structure, which is exported to the exporters through `LogExporter::emit(logdata)`. This ensures that resource is not sent to the exporter with each event.
~~2. Avoid cloning LogRocord during emit if single processor is configured.~~ Will create separate PR for this

call sequence -
`LoggerProvider::build()` -> `LogProcessor::set_resource()` -> `LogExporter::set_resource()`

Currently, these changes apply exclusively to Logs. Once there is consensus and these changes are merged, will extend them to traces and metrics. Additionally, changes made in #1526 can be undone after implementing this across all signals.

**Benchmark**:

***Main branch***:
```
Number of threads: 8
Throughput: 16,356,200 iterations/sec
Throughput: 15,894,400 iterations/sec
Throughput: 16,018,200 iterations/sec
Throughput: 16,344,800 iterations/sec
```

***PR branch***: (with Resource propagation at startup )
```
Number of threads: 8
Throughput: 19,301,400 iterations/sec
Throughput: 19,789,000 iterations/sec
Throughput: 20,133,600 iterations/sec
Throughput: 19,763,000 iterations/sec
```

~~***PR branch***: (with Resource propagation at startup + avoid LogRecord copy for single processor)~~ will create separate PR for this
```
Number of threads: 8
Throughput: 28,997,800 iterations/sec
Throughput: 29,057,800 iterations/sec
Throughput: 28,755,800 iterations/sec
Throughput: 29,712,000 iterations/sec
```

## Merge requirement checklist

* [x] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-rust/blob/main/CONTRIBUTING.md) guidelines followed
* [x] Unit tests added/updated (if applicable)
* [x] Appropriate `CHANGELOG.md` files updated for non-trivial, user-facing changes
* [x] Changes in public API reviewed (if applicable)

### CODE REVIEW

File: opentelemetry-otlp/src/logs.rs
Comment (cijothomas) [inline opentelemetry-otlp/src/logs.rs:108]: is this meant to be fixed in this PR itself?

File: opentelemetry-sdk/CHANGELOG.md
Comment (cijothomas) [inline opentelemetry-sdk/CHANGELOG.md:17]:
```suggestion
    - Add `set_resource` method in LogProcessor and Exporter traits
```

File: opentelemetry-sdk/src/logs/log_processor.rs
Comment (cijothomas) [inline opentelemetry-sdk/src/logs/log_processor.rs:57]: I suggest to provide an empty implementation here. Not all processor author will be interested in doing anything with Resource, so its best to make it optional.

File: opentelemetry-sdk/src/export/logs/mod.rs
Comment (cijothomas) [inline opentelemetry-sdk/src/export/logs/mod.rs:25]: same comment as https://github.com/open-telemetry/opentelemetry-rust/pull/1636/files#r1543481098.

File: opentelemetry-sdk/src/logs/log_processor.rs
Comment (cijothomas) [inline opentelemetry-sdk/src/logs/log_processor.rs:167]: could we clone the resource, and then hand it over to the BatchMessage::SetResource? Is Arc needed? The processor need not keep ownership, so it should be okay to move ownership to BatchMessage?

**Comment by cijothomas** [comment]
Looks good overall! Thank you.
Will wait couple days to get more eyes on this approach!

**Comment by lalitb** [comment]
> @lalitb any noticeable change in performance, out of curiosity!

Yes, there is a visible improvement with stress-test in the hot path as we don't clone and add resources in the LogData during emit. There should be improvement in memory usage too, but we don't have a direct way of measuring it.

*with PR*
```console
$ cargo run --bin logs --all-features --release
   Compiling opentelemetry_sdk v0.22.1 (/home/labhas/otel/rust/lalit/set-resource/opentelemetry-rust/opentelemetry-sdk)
   Compiling stress v0.1.0 (/home/labhas/otel/rust/lalit/set-resource/opentelemetry-rust/stress)
    Finished `release` profile [optimized] target(s) in 4.21s
     Running `target/release/logs`
Number of threads: 8
Throughput: 13,469,200 iterations/sec
Throughput: 13,271,600 iterations/sec
Throughput: 13,911,600 iterations/sec
```

*main branch*
```console
labhas@CPC-labha-5U0JP:~/otel/rust/lalit/set-resource/opentelemetry-rust$ cargo run --bin logs --all-features --release
   Compiling opentelemetry_sdk v0.22.1 (/home/labhas/otel/rust/lalit/set-resource/opentelemetry-rust/opentelemetry-sdk)
   Compiling opentelemetry-appender-tracing v0.3.0 (/home/labhas/otel/rust/lalit/set-resource/opentelemetry-rust/opentelemetry-appender-tracing)
   Compiling stress v0.1.0 (/home/labhas/otel/rust/lalit/set-resource/opentelemetry-rust/stress)
    Finished `release` profile [optimized] target(s) in 4.22s
     Running `target/release/logs`
Number of threads: 8
Throughput: 11,924,800 iterations/sec
Throughput: 11,722,800 iterations/sec
Throughput: 11,945,000 iterations/sec
```

File: opentelemetry-sdk/src/logs/log_processor.rs
Comment (lalitb) [reply opentelemetry-sdk/src/logs/log_processor.rs:167]: Yes Arc is needed for batch-processor, to share the resource across separate thread. I believe it should be fine to have one-time overhead during startup.

File: opentelemetry-sdk/src/export/logs/mod.rs
Comment (lalitb) [reply opentelemetry-sdk/src/export/logs/mod.rs]: Done.

File: opentelemetry-sdk/src/logs/log_processor.rs
Comment (lalitb) [reply opentelemetry-sdk/src/logs/log_processor.rs]: Good suggestion. Done.

File: opentelemetry-otlp/src/logs.rs
Comment (lalitb) [reply opentelemetry-otlp/src/logs.rs:108]: Yes, it is removed now. We are relying on the default resources.

Comment (lalitb) [reply opentelemetry-otlp/src/logs.rs]: Good catch, the change was leftover in my local. It is required, otherwise OTLP log exporter won't get the resource. I think we need unit-test for OTLP log exporter validate this. Let me see if I can add in this PR.

File: opentelemetry-sdk/benches/log.rs
Comment (cijothomas) [inline opentelemetry-sdk/benches/log.rs:24]: we need this?

File: opentelemetry-sdk/src/logs/log_emitter.rs
Comment (cijothomas) [inline opentelemetry-sdk/src/logs/log_emitter.rs:369]: we can remove this now, right?

File: opentelemetry-sdk/src/testing/logs/in_memory_exporter.rs
Comment (cijothomas) [inline opentelemetry-sdk/src/testing/logs/in_memory_exporter.rs:127]: maybe we can keep the existing public api, and add a new api for get_resource.

Comment (lalitb) [reply opentelemetry-sdk/src/testing/logs/in_memory_exporter.rs:127]: The get_emitted_logs() still returns the data in the same (older LogData) format, only the struct name is now `LogDataWithResource` instead of `LogData`.

File: opentelemetry-sdk/CHANGELOG.md
Comment (lalitb) [reply opentelemetry-sdk/CHANGELOG.md]: Good point, This is modified now:
- [#1636](https://github.com/open-telemetry/opentelemetry-rust/pull/1636) [Logs SDK] Improves performance by sending 
  Resource information to processors (and exporters) once, instead of sending with every log. If you are an author
  of Processor, Exporter, the following are *BREAKING* changes.
    - Implement `set_resource` method in your custom LogProcessor, which invokes exporter's `set_resource`.
    - Implement `set_resource` method in your custom LogExporter. This method should save the resource object
      in original or serialized format, to be merged with every log event during export.
    - `LogData` doesn't have the resource attributes. The `LogExporter::export()` method needs to merge it 
      with the earlier preserved resource before export.

**Comment by lalitb** [comment]
Have update this PR to not clone `LogRecord` within `Logger:emit()` if only single processor is configured - I believe this would be the common use-case. And can see vast improvement in the stress test. Have updated the latest stresss test in the PR description.

File: opentelemetry-sdk/src/logs/log_emitter.rs
Comment (TommyCpp) [inline opentelemetry-sdk/src/logs/log_emitter.rs:244]: Does this need to pass mutable reference so that the one processor results can be pass to next processor?

Comment (lalitb) [reply opentelemetry-sdk/src/logs/log_emitter.rs]: have removed the single processor optimization. Will create separate PR for that along, and also if we can use mutable reference.

**Comment by lalitb** [comment]
Seems #1585 has some merge conflicts, fixing them now.

**Comment by lalitb** [comment]
> Seems merge conflicts with #1585, fixing them now.

should be good now.

File: stress/src/logs.rs
Comment (cijothomas) [inline stress/src/logs.rs:31]: we don't need this, right?

**Review by cijothomas — APPROVED** [review]

File: stress/src/logs.rs
Comment (lalitb) [reply stress/src/logs.rs]: Ah yes. Removed now.

**Comment by lalitb** [comment]
Please don't merge yet. OTLP export seems to be broken after merging changes with #1585. Looking into it

**Comment by lalitb** [comment]
> Please don't merge yet. OTLP http log export seems to be broken after merging changes with #1585. Looking into it

Should be good now. Need to add integration tests for OTLP Log exporter to detect these issues. Will create a PR for this.

**Pull request merged**
